### ISSUE TITLE
Fix bug causing garbled text rendering when multiple maps are rendered on the page

### ISSUE THREAD

**Issue by christianbauer** [mapbox/mapbox-gl-js#3086]
**0.23.0**

 I've upgraded from 0.18.0.

This is a larger GWT application that uses JsInterop, so it's difficult for me to provide code to run. Before I isolate and try to reproduce this error with plain JS, it would be good to know if this is a known issue.

Ultimately all I do is use several instance of Map on the same DOM, and attach/detach them frequently. When the first instance of Map renders, it looks fine. When I detach the first instance and create and attach a second (and probably other) instance(s) of a Map, this second instance has broken rendering of text labels.

See the attached screenshot.

The code is unchanged except for some necessary API adjustments with handling GeoJSON sources.
<img width="846" alt="screen shot 2016-08-27 at 22 13 24" src="https://cloud.githubusercontent.com/assets/1006685/18029856/5a4dcb66-6ca4-11e6-8b79-9452d1ed4ff0.png">

**Comment by mourner** [comment]
Can you set up a minimal JSFiddle that reproduces the case? cc @bhousel

**Comment by christianbauer** [comment]
https://jsfiddle.net/bL104L7k/4/

Click the Toggle Map button several times, it takes me between one and five clicks to see the error and it reverts back to normal in between. The width of the map container, or the fact that some part is not visible in the viewport initially seems to make a difference.

I couldn't see the error when I switched to version 0.22.0.

Chrome Version 52.0.2743.116 (64-bit)
OS X 10.11.6

**Comment by lucaswoj** [comment]
This bug is reminiciant of and perhaps shares a smiliar root cause to https://github.com/mapbox/mapbox-gl-js/issues/2012

**Comment by lucaswoj** [comment]
@joedjc can you share your degugging code from https://github.com/mapbox/mapbox-gl-js/issues/3126 here?

**Comment by jfirebaugh** [comment]
In #2952, where's the fan out from the worker pool to the appropriate map-specific receiver for messages transmitted from worker to main such as `'get sprite json'`, and `'get glyphs'`? Unless I'm missing something it looks like every instantiated `Style` object will receive and respond to these messages from workers working on behalf of any map.

cc @anandthakker

**Comment by anandthakker** [comment]
@jfirebaugh that's right.  found this and included a patch over in #2982, but the fix is in a single commit ( https://github.com/mapbox/mapbox-gl-js/pull/2982/commits/8a7807f6f7785ee0c54b763aa5c765b4ea807451 ) so would be easy to pull into a separate PR if that's helpful

**Comment by YunjieLi** [comment]
Hey just wanted to let you know I ran into the same issue again in Studio last night. 
I had several maps open in different tabs at the time. In the first style it was in the editing page and went away only after I previewed the style. I ran into it again in the publishing window of a second style I created later last night but not in editing view. 

![image](https://cloud.githubusercontent.com/assets/12749701/18643044/27fbc328-7e58-11e6-886f-349dd4a61d36.png)

**Comment by lucaswoj** [comment]
@YunjieLi this fix has not made it into a release yet

**Comment by karlguillotte** [comment]
Confirm it is now working! Thanks for releasing.

**Comment by mboeringa** [comment]
Hi,

I just wanted to add that I saw a similar issue yesterday on Mapbox Streets main demo page, as viewed in Internet Explorer 11 on Windows 7. What happened is that **when zooming in and out**, the labels of some features got garbled, but then, after maybe a second or so, they got re-rendered as proper labels.

Notice the double labels in the attached image. As said, this condition only lasted for a second or so, after which the city name labels like "Leeuwarden" and "Groningen", were properly displayed without the secondary garbled text.

![garbled_text](https://cloud.githubusercontent.com/assets/7635726/20455749/78cb969e-ae63-11e6-8c4d-7187e19f806a.jpg)

**Comment by lucaswoj** [comment]
Thanks for the report @mboeringa. I cut a fresh ticket at #3669

**Comment by mboeringa** [comment]
Ok, good to hear.
